apps/studio/src/components/sidebar/core/table_list/VirtualTableList.vue

```diff
@@ -0,0 +1,195 @@
+<template>
+  <virtual-list
+    ref="vList"
+    class="list-body"
+    :data-key="'key'"
+    :data-sources="displayItems"
+    :data-component="itemComponent"
+    :estimate-size="itemHeight"
+    :keeps="keeps"
+    :extra-props="{ onExpand: handleExpand, onPin: handlePin }"
+  ></virtual-list>
+</template>
+
+<script lang="ts">
+import Vue from "vue";
+import { TableOrView, Routine } from "@/lib/db/models";
+import TableListContextMenus from "@/mixins/TableListContextMenus";
+import ItemComponent from "./Item.vue";
+import VirtualList from "vue-virtual-scroll-list";
+import { AppEvent } from "@/common/AppEvent";
+import { mapGetters, mapState } from "vuex";
+import { PinnedEntity } from "@/common/appdb/models/PinnedEntity";
+
+type Entity = TableOrView | Routine | string;
+
+type Item = SchemaItem | TableItem;
+
+interface BaseItem {
+  type: "schema" | "table" | "routine";
+  entity: Entity;
+  key: string;
+  expanded: boolean;
+  hidden: boolean;
+  contextMenu: any[];
+  level: number;
+}
+
+interface SchemaItem extends BaseItem {
+  type: "schema";
+  entity: string;
+  skipDisplay: boolean;
+}
+
+interface TableItem extends BaseItem {
+  type: "table" | "routine";
+  entity: TableOrView | Routine;
+  parent: SchemaItem;
+  pinned: boolean;
+}
+
+export default Vue.extend({
+  mixins: [TableListContextMenus],
+  components: { VirtualList },
+  data() {
+    return {
+      items: [],
+      displayItems: [],
+      itemComponent: ItemComponent,
+      itemHeight: 22.8,
+      keeps: 30,
+    };
+  },
+  methods: {
+    generateItems() {
+      const items = [] as Item[];
+      const noFolder = this.schemaTables.length === 1;
+      this.schemaTables.forEach((schema: any) => {
+        const schemaItem: SchemaItem = {
+          type: "schema",
+          key: schema.schema,
+          entity: schema.schema,
+          expanded: true,
+          hidden: this.hiddenSchemas.includes(schema.schema),
+          contextMenu: this.schemaMenuOptions,
+          skipDisplay: noFolder,
+          level: 0,
+        };
+        items.push(schemaItem);
+        schema.tables.forEach((table: TableOrView) => {
+          items.push({
+            type: "table",
+            key: `${schema.schema}.${table.name}.table`,
```

Comment:
Feels like we should have an `entityId` helper in `utils` so we could call `entityId(schema, table, 'table')`